sycl/source/detail/kernel_impl.hpp

```diff
@@ -162,6 +162,39 @@ class kernel_impl {
   ext_oneapi_get_info(queue Queue, const range<3> &MaxWorkGroupSize,
                       size_t DynamicLocalMemorySize) const;
 
+  /// Query queue/launch-specific information from a kernel using the
+  /// info::kernel_queue_specific descriptor for a specific Queue and values.
+  /// max_num_work_groups is the only valid descriptor for this function.
+  ///
+  /// \param Queue is a valid SYCL queue.
+  /// \param WG is a work group size
+  /// \return depends on information being queried.
+  template <typename Param>
```

Comment:
Should we perhaps use some kind of SFINAE here to explicitly check that `Param` has the value defined in the spec?

It seems like at the moment we are doing SFINAE only on the return type by saying `Param::return_type` but there could be other descriptors that will have this type defined and we will end up instantiating declarations for them if the user supplies an invalid Param by mistake. This could lead to some weird linker errors or some runtime bug.

Reply:
This isn't user-visible API. Don't we have proper constraints there?

Reply:
The user-visible API in `kernel.hpp` which calls this one also does something similar with `typename detail::is_kernel_queue_specific_info_desc<Param>::return_type` which again, could potentially allow other `Param` types to slip through the cracks. Ideally, we'd like to follow the spec to the word and mandate `Param` to be as required by the spec through SFINAE.